Automatically rerun nightly builds which fail due to transient connection errors

andrewmogan opened the pull request:
Addresses #422. The primary retry logic lives in `scripts/spack/build-release*.sh`. If the `spack install` command fails for any reason, the build scripts will now scan a log file for a spack `FetchError`, in which case it will rerun the install command up to two additional times before giving up. If the build fails for any reason that isn't a `FetchError`, it will simply exit with the `spack install` exit code. 

[This test build](https://github.com/DUNE-DAQ/daq-release/actions/runs/12791004951) shows the results of simulating two consecutive `FetchError`s, followed by a successful build. And [here](https://github.com/DUNE-DAQ/daq-release/actions/runs/12793271863) is the same test performed using the v4 workflow. 

This also adds a step in the `build_the_develop_release_spack` job which sets the environment variables `NIGHTLY_TAG`, `BASE_RELEASE_DIR`, `DET_RELEASE_DIR`, and `OS` for use in both the `coredaq` and `fddaq` installs. This was originally due to the inability to set these environment variables using an external retry action, but I decided to keep this step even after giving up on said action since it reduces code duplication.

andrewmogan committed:
Replace spack install with intentional failure and raise specific exit code

andrewmogan committed:
Issue #422: Use retry action on coredaq build step of v5 nightly

andrewmogan committed:
Issue #422: Arbitrarily set (required) retry timeout

andrewmogan committed:
Set environment variables in a separate step

This is because, as far as I can tell, the retry action doesn't allow
you to do this in the command field.

andrewmogan committed:
Issue #422: Checkout this feature branch of daq-release for testing

andrewmogan committed:
Issue #422: Close if statements

andrewmogan committed:
Issue #422: Add file argument to grep

andrewmogan committed:
Issue #422: Specify bash shell in retry action

andrewmogan committed:
Issue #422: Add debug echo statements

andrewmogan committed:
Issue #422: Add echo debug statements

andrewmogan committed:
Issue #422: Propagate exit 111 to github actions step

andrewmogan committed:
Issue #422: Set exist_ok to True for generate_daq_package

andrewmogan committed:
Issue #422: Set exist_ok to True everywhere

andrewmogan committed:
Issue #422: Revert exist_ok=True's and set on_retry_command for retry action

andrewmogan committed:
Issue #422: Sync coredaq and fddaq install steps

andrewmogan committed:
Issue #422: Remove debug statements

andrewmogan committed:
Remove intentional failure

andrewmogan committed:
Issue #422: Reimplement full spack install

andrewmogan committed:
Issue #422: Replace retry action with manual logic

andrewmogan committed:
Revert to forced exit 111 for testing

andrewmogan committed:
Issue #422: echo spack install exit code

andrewmogan committed:
Issue #422: Cleanup directories before installing

andrewmogan committed:
Issue #422: Attempt to spack rm a problematic directory

andrewmogan committed:
Issue #422: Move spack rm command

andrewmogan committed:
Issue #422: Move spack repo rm to build-release.sh

andrewmogan committed:
Issue #422: Fix release manifest path

andrewmogan committed:
Issue #422: Use exit 111 for simulated failure

andrewmogan committed:
Issue #422: Revert to previous exit code and debug statements

andrewmogan committed:
Issue #422: Clean build directories before spack install

andrewmogan committed:
Issue #422: Fix build-release argument

andrewmogan committed:
Issue #422: Move and narrow cleanup commands

andrewmogan committed:
Issue #422: Remove spack-repo directory on exit 111

andrewmogan committed:
Issue #422: Remove spack default directory on retry

andrewmogan committed:
Issue #422: Move rm commands to workflow rather than build-release.sh

andrewmogan committed:
Issue #422: Remove local spack area

andrewmogan committed:
Issue #422: Remove rm statements back to build-release and adjust retry logic

andrewmogan committed:
Issue #422: Remove spack version area on exit 111

andrewmogan committed:
Issue #422: Add debug echo statements

andrewmogan committed:
Issue #422: Move retry logic entirely to build-release.sh

andrewmogan committed:
Issue #422: Improve retry logic and logging

andrewmogan committed:
Issue #422: Run on protodune-daq01

andrewmogan committed:
Issue #422: Intentionally fail retries

andrewmogan committed:
Issue #422: Fail first, then spack install on retry

andrewmogan committed:
Issue #422: Run on daq

andrewmogan committed:
Issue #422: Keep spack version directory and revert to intentional failure

andrewmogan committed:
Issue #422: Revert to full spack install on first retry

andrewmogan committed:
Issue #422: Revert back to intentional failure on retry and remove spack-repo directory

andrewmogan committed:
Issue #422: Revert back to full install on first retry

andrewmogan committed:
Issue #422: Remove deprecated if statement

andrewmogan committed:
Issue #422: ls SPACK_AREA for debugging

andrewmogan committed:
Issue #422: ls deeper for more debugging

andrewmogan committed:
Issue #422: More ls

andrewmogan committed:
Issue #422: Remove more specific spack directories and attempt another full install after retry

andrewmogan committed:
Issue #422: Add more ls and recursive diff for debugging

andrewmogan committed:
Issue #422: Don't rm anything, see what happens

andrewmogan committed:
Issue #422: Fix coredaq artifact upload arguments

andrewmogan committed:
Issue #422: Intentionally fail each retry

andrewmogan committed:
Issue #422: Run full spack install each time, but still force failure

andrewmogan committed:
Issue #422: Fail the first two builds, then succeed

andrewmogan committed:
Issue #422: Clean up: remove debug statements, comments, etc.

andrewmogan committed:
Issue #422: Add retry logic with intentional failure to v4 build

andrewmogan committed:
Issue #422: Remove intentional failures

andrewmogan committed:
Issue #422: Prepare for merge by checking out develop of daq-release instead of a feature branch

Commit:
Merge branch 'develop' into amogan/issue422_scan_for_fetcherrors

jcfreeman2 commented:
Thanks for doing this, and apologies for the long wait before a review. My comments:
* I merged the `develop` branch into this feature branch; the only effect this had on what you wrote is that the hardwired `gcc` version got replaced by a `GCC_VERSION` environment variable
* I love that the variables in the nightly workflows (`BASE_RELEASE_DIR`, etc.) now no longer violate [Don't Repeat Yourself]( https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
* It _looks_ like if the `spack install` of `fddaq` results in an error which isn't a connection issue, the Workflow continues?
* Getting back to D.R.Y.,  rather than run `spack install...`, checking the result, and then starting a loop over `spack install...` if there was a hangup, it would be a bit less verbose if you just put the call in a loop to begin with. In pseudocode, the body of the loop would be:
```
spack install fddaq
Check result. 
If it's fine, break out of the loop
If it's a connection error, redo the loop unless we've hit max tries
If it's another kind of error, end the Workflow
```
[this page](https://stackoverflow.com/questions/24420342/is-there-a-do-while-loop-in-bash) explains how to do a construct like this in shell scripting
* The same "connection protection" logic should apply to `dbe` as well as to `fddaq`

andrewmogan committed:
Simplify retry logic into a single do-while loop

andrewmogan committed:
Issue #422: Extend retry logic to dbe

andrewmogan committed:
Simulate fetch error on first build attempt

andrewmogan committed:
Fix spack install exit code line

andrewmogan committed:
Correct intentional failure on first try, maybe

andrewmogan committed:
Issue #422: Intentionally fail every build attempt

andrewmogan committed:
Issue #422: Make sure fddaq step doesn't continue on failure

andrewmogan committed:
Issue #422: Remove intentional failure

andrewmogan committed:
Merge branch 'develop' of https://github.com/DUNE-DAQ/daq-release into amogan/issue422_scan_for_fetcherrors

andrewmogan committed:
Issue #422: Sync retry logic for v4 build

andrewmogan committed:
Issue #422: Simulate non-retryable exit code on fddaq step

andrewmogan committed:
Fix syntax for simulated failures

andrewmogan committed:
Correctly update spack_install_exit_code for simulated failure

andrewmogan committed:
Issue #422: Revert intentional failures

andrewmogan commented:
I've finished addressing the above comments. The retry logic is now all contained in a `while true; do` loop, and applies to `dbe` as well. Incidentally, I removed the `is_connection_error` boolean, which should remove the possibility of continuing after the fddaq step if it fails due to a non-connection error. Test builds are listed below. Note the searching the GitHub output for "attempt number" will bring you to the relevant lines.

[Simulate a connection error during first coredaq attempt, followed by non-connection failure in fddaq](https://github.com/DUNE-DAQ/daq-release/actions/runs/12954623561/job/36136961893)

[Simulate connection error on the first spack install attempt](https://github.com/DUNE-DAQ/daq-release/actions/runs/12937924200/job/36086946921) 

[Fail with connection error, but only on fddaq step](https://github.com/DUNE-DAQ/daq-release/actions/runs/12953118490/job/36132051287)

Commit:
JCF: Issue #422: add attempt reporting to dbe alongside coredaq/fddaq reporting; alsa opportunistically drop dbe-build logic from the v4 nightly build since dbe isn't a thing in v4

jcfreeman2 commented:
Looks good, approved. Note that I added a commit on this feature branch which adds the helpful attempt-reporting scheme to the `dbe` build (it was already there for `coredaq` and `fddaq`) in `build-release.sh`; I also noticed that we don't need to even attempt building`dbe` in `build-release-v4.sh` so I removed that part of the script.
